netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -813,116 +852,239 @@ _net_ebpf_extension_connection_context_initialize(
+    _net_ebpf_ext_sock_addr_lock) static void _net_ebpf_ext_purge_connect_contexts_under_lock(BOOLEAN delete_all)
 {
     uint64_t expiry_time = CONVERT_100NS_UNITS_TO_MS(KeQueryInterruptTime()) - EXPIRY_TIME;
 
-    LIST_ENTRY* list_entry = _net_ebpf_ext_connect_context_list.Blink;
-    while (list_entry != &_net_ebpf_ext_connect_context_list) {
+    // Free entries from the LRU list. These entries should also be removed from the hash table.
+    LIST_ENTRY* list_entry = _net_ebpf_ext_blocked_connect_context_lru_list.Blink;
+    while (list_entry != &_net_ebpf_ext_blocked_connect_context_lru_list) {
         net_ebpf_extension_connection_context_t* entry =
             CONTAINING_RECORD(list_entry, net_ebpf_extension_connection_context_t, list_entry);
         if (!delete_all && entry->timestamp > expiry_time) {
             break;
         }
 
 #pragma warning(suppress : 6001) /* entry and list entry are non-null */
-        if (ebpf_hash_table_delete(_net_ebpf_ext_connect_context_hash_table, (uint8_t*)entry) != EBPF_SUCCESS) {
+        if (ebpf_hash_table_delete(_net_ebpf_ext_blocked_connect_context_hash_table, (uint8_t*)entry) != EBPF_SUCCESS) {
             // If we fail to delete it, leave it in the list so that we can try again later
             continue;
         }
 
         list_entry = list_entry->Blink;
         RemoveEntryList(&entry->list_entry);
 
-        _net_ebpf_ext_connect_context_count--;
+        --_net_ebpf_ext_connect_context_count;
 
         NET_EBPF_EXT_LOG_MESSAGE_UINT64(
             NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
             NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
-            "_net_ebpf_ext_purge_lru_contexts_under_lock: Delete",
+            "_net_ebpf_ext_purge_connect_contexts_under_lock: Delete",
             entry->transport_endpoint_handle);
 
         ExFreePool(entry);
     }
 
+    // Free entries from low-memory list. These entries should be returned to the free list.
+    list_entry = _net_ebpf_ext_low_memory_blocked_connect_context_list.Blink;
+    while (list_entry != &_net_ebpf_ext_low_memory_blocked_connect_context_list) {
+        net_ebpf_extension_connection_context_t* entry =
+            CONTAINING_RECORD(list_entry, net_ebpf_extension_connection_context_t, list_entry);
+        if (!delete_all && entry->timestamp > expiry_time) {
+            break;
+        }
+
+#pragma warning(suppress : 6001) /* entry and list entry are non-null */
+        list_entry = list_entry->Blink;
+        RemoveEntryList(&entry->list_entry);
+
+        if (delete_all) {
+            // Free the memory
+            ExFreePool(entry);
+        } else {
+            // Return the entry to the free list
+            InsertHeadList(&_net_ebpf_ext_low_memory_free_connect_context_list, &entry->list_entry);
+        }
+    }
+
+    // Free pre-allocated entries if we are shutting down.
+    if (delete_all) {
+        while (!IsListEmpty(&_net_ebpf_ext_low_memory_free_connect_context_list)) {
```

Comment:
Do we need this extra loop, since we are already deleting the entries in the above loop if `delete_all` is true?

Reply:
I have moved this to the newly created cleanup function. We have two lists, one for the not used pre-allocated contexts, and then one for the in-use contexts. This first part would move or cleanup in-use entries, then the cleanup function logic goes through the not in-use contexts and cleans those up.